### PR TITLE
Report progress in long-running functions

### LINKED ISSUE #249: Implement progress bar for long-running functions
From https://github.com/openjournals/joss-reviews/issues/3073#issuecomment-817023969:

> When testing preprocess_data() and run_ml() on a larger dataset with 1917 rows × 14426 columns, it would take an indeterminately long time. It should have a progress bar indicating the iteration #, and the estimated duration for each iteration, similar to Python's tqdm. 

And https://github.com/openjournals/joss-reviews/issues/3073#issuecomment-828720647:

> I agree with @JonnyTran 's suggestion to display time to completion estimates and relevant metrics (e.g. loss or chosen metrics for current iteration) during training. The average user will find this helpful. This would also be the case when parallelizing `run_ml()` execution.

I'll probably use https://github.com/r-lib/progress and make it optional.

### PR DESCRIPTION
**Pull request by kelly-sovacool**
## Issues
- Resolves #249.

## Change(s) made
- Implement an optional progress bar to report the elapsed time & estimated time to completion for:
  - [x] `preprocess_data()`
  - [ ] ~`run_ml()`~  – see https://github.com/SchlossLab/mikropml/issues/249#issuecomment-832259941
  - [x] `get_feature_importance()`
- New internal helper function:
  - `pbtick()` - update the progress bar if it's not `NULL`.
- New suggested (optional) dependencies: [`progressr`](https://progressr.futureverse.org/), [`progress`](https://github.com/r-lib/progress)

I used the `progressr` package because it supports progress bars in `future.apply` functions. How to report the progress is up to the user. Example usage:

```r
# optionally, specify the progress bar format with the `progress` package
progressr::handlers(progressr::handler_progress(
    format = ":message :bar :percent | elapsed: :elapsed | eta: :eta",
    clear = FALSE,
    show_after = 0))
# tell progressr to always report progress in any functions that use it
progressr::handlers(global = TRUE)

# run your code and watch the live progress updates
library(mikropml)
dat <- preprocess_data(otu_mini_bin, 'dx')$dat_transformed
#> Using 'dx' as the outcome column.                                            
#> preprocessing ========================>-------  78% | elapsed:  1s | eta:  0s
results <- run_ml(dat, "glmnet", kfold = 2, cv_times = 2, 
                  find_feature_importance = TRUE)
#> Using 'dx' as the outcome column.
#> Training the model...
#> Training complete.
#> Feature importance =========================== 100% | elapsed: 37s | eta:  0s
```

If a user calls these mikropml functions without either setting `progressr::handlers(global = TRUE)` or wrapping the code in `progressr::with_progress()`, the progress will not be reported.

### Limitations

- I tried to avoid calling `pbtick()` inside every loop/*apply/map iteration in `preprocess_data()` because it considerably slows down performance when I tested it with `data-raw/otu_large_bin.csv`. As a result, the resolution is somewhat coarse, but the ETA still appears to be fairly accurate. 
- Sometimes `preprocess_data()` finishes before the progress bar reaches 100%, i.e. the total number of steps is sometimes too large depending on the dataset used. I may circle back to this later to try to make it more precise.

## Checklist

(~Strikethrough~ any points that are not applicable.)

- [x] Write unit tests for any new functionality.
- [x] Update roxygen comments & vignettes if there are any API changes.
  - [x] roxygen docs
  - [x] parallel vignette 
- [x] Update `NEWS.md` if this includes any user-facing changes. 
- [x] The check workflow succeeds on your most recent commit. **This is always required before the PR can be merged.**

### CODE REVIEW

File: vignettes/parallel.Rmd [hunk 1]
```diff
@@ -144,6 +145,34 @@ perf_boxplot +
...
+#> preprocessing ========================>-------  78% | elapsed:  1s | eta:  0s
+results <- run_ml(dat, "glmnet", kfold = 2, cv_times = 2,
+                  find_feature_importance = TRUE)
+#> Using 'dx' as the outcome column.
+#> Training the model...
+#> Training complete.
+#> Feature importance =========================== 100% | elapsed: 37s | eta:  0s
+```
+
+
```
Comment (zenalapp) [inline vignettes/parallel.Rmd:175]: Thoughts on adding a line that says how to remove the progress bar if you decide you don't want it?

```
progressr::handlers(global = FALSE)
```

File: tests/testthat/test-utils.R [hunk 2]
```diff
@@ -70,3 +70,11 @@ test_that("replace_spaces works", {
...
   )
 })
+
+test_that("pbtick() updates the progress bar", {
+  f <- function() {
+    pb <- progressr::progressor(steps = 5)
+    pbtick(pb, message = "progress!")
+  }
+  expect_condition(expect_invisible(f()))
+})
```
Comment (zenalapp) [inline tests/testthat/test-utils.R:80]: I noticed our code coverage is lower for utils now. Maybe add more unit tests for when pb and/or message are `NULL`?

**Review by zenalapp — COMMENTED** [review]
Works perfectly! Just left a few suggestions - you can decide if you think it's worth addressing them.

File: tests/testthat/test-utils.R (hunk 2 above)
Comment (kelly-sovacool) [reply tests/testthat/test-utils.R:80]: That codecov report is slightly outdated (deleted it now). After [this commit](https://github.com/SchlossLab/mikropml/pull/257/commits/0b8299103add44cbe88e7dcebbac4f6099563088) the coverage for utils is 100%.

**Review by sklucas — APPROVED** [review]
A nice additional functionality that I think users will really appreciate.

File: vignettes/parallel.Rmd (hunk 1 above)
Comment (kelly-sovacool) [reply vignettes/parallel.Rmd:182]: I encourage them to read the `progressr` docs instead: https://github.com/SchlossLab/mikropml/pull/257/commits/9e2dc5f1e225c0556a5edf0432083d9fffdade1d

**Review by zenalapp — APPROVED** [review]

**Comment by kelly-sovacool** [comment]
I'm rebuilding the `parallel` vignette locally (it takes a while), then will merge this.
